### ISSUE TITLE
Bug on android related to shaders

### ISSUE THREAD

**Issue by Neizvestnyj** [kivymd/KivyMD#1352]
**Full log**
```
09-18 01:41:14.903 21996 22024 I python  : AND: Ran string
09-18 01:41:14.903 21996 22024 I python  : Run user program, change dir and execute entrypoint
09-18 01:41:15.019 21996 22024 I python  : [INFO   ] [Logger      ] Record log in /data/user/0/<APP>/files/app/.kivy/logs/kivy_22-09-18_1.txt
09-18 01:41:15.019 21996 22024 I python  : [INFO   ] [Kivy        ] v2.1.0
09-18 01:41:15.019 21996 22024 I python  : [INFO   ] [Kivy        ] Installed at "/data/user/0/<APP>/files/app/_python_bundle/site-packages/kivy/__init__.pyc"
09-18 01:41:15.019 21996 22024 I python  : [INFO   ] [Python      ] v3.9.9 (main, Sep 18 2022, 01:31:35) 
09-18 01:41:15.019 21996 22024 I python  : [Clang 14.0.6 (https://android.googlesource.com/toolchain/llvm-project 4c603efb
09-18 01:41:15.020 21996 22024 I python  : [INFO   ] [Python      ] Interpreter at ""
09-18 01:41:15.020 21996 22024 I python  : [INFO   ] [Logger      ] Purge log fired. Processing...
09-18 01:41:15.021 21996 22024 I python  : [INFO   ] [Logger      ] Purge finished!
09-18 01:41:15.714 21996 22024 I python  : [INFO   ] [Factory     ] 189 symbols loaded
09-18 01:41:15.959 21996 22024 I python  : [INFO   ] [Image       ] Providers: img_tex, img_dds, img_sdl2, img_pil (img_ffpyplayer ignored)
09-18 01:41:15.989 21996 22024 I python  : [INFO   ] [Window      ] Provider: sdl2
09-18 01:41:16.008 21996 22024 I python  : [INFO   ] [GL          ] Using the "OpenGL ES 2" graphics system
09-18 01:41:16.009 21996 22024 I python  : [INFO   ] [GL          ] Backend used <sdl2>
09-18 01:41:16.009 21996 22024 I python  : [INFO   ] [GL          ] OpenGL version <b'OpenGL ES 3.2 V@415.0 (GIT@033a5b0, I0e419467bc, 1583931100) (Date:03/11/20)'>
09-18 01:41:16.010 21996 22024 I python  : [INFO   ] [GL          ] OpenGL vendor <b'Qualcomm'>
09-18 01:41:16.010 21996 22024 I python  : [INFO   ] [GL          ] OpenGL renderer <b'Adreno (TM) 630'>
09-18 01:41:16.010 21996 22024 I python  : [INFO   ] [GL          ] OpenGL parsed version: 3, 2
09-18 01:41:16.010 21996 22024 I python  : [INFO   ] [GL          ] Texture max size <16384>
09-18 01:41:16.011 21996 22024 I python  : [INFO   ] [GL          ] Texture max units <16>
09-18 01:41:16.031 21996 22024 I python  : [INFO   ] [Window      ] auto add sdl2 input provider
09-18 01:41:16.032 21996 22024 I python  : [INFO   ] [Window      ] virtual keyboard not allowed, single mode, not docked
09-18 01:41:16.034 21996 22024 I python  : [INFO   ] [KivyMD      ] 1.1.0.dev0, git-Unknown, 2022-09-17 (installed at "/data/user/0/<APP>/files/app/_python_bundle/site-packages/kivymd/__init__.pyc")
09-18 01:41:16.040 21996 22024 I python  : [INFO   ] [Text        ] Provider: sdl2
09-18 01:41:16.311 21996 22024 I python  : [INFO   ] [GL          ] NPOT texture support is available
09-18 01:41:16.332 21996 22024 I python  : [WARNING] [Base        ] Unknown <android> provider
09-18 01:41:16.333 21996 22024 I python  : [INFO   ] [Base        ] Start application main loop
09-18 01:41:16.601 22032 22032 F DEBUG   :       #06 pc 00006803  /data/data/<APP>/files/app/_python_bundle/site-packages/kivy/graphics/vbo.so
```

After android app start I get **error**:
```
09-18 00:04:30.928 30527 30527 F DEBUG   :       #04 pc 0000000000009e40  /data/data/<APP>/files/app/_python_bundle/site-packages/kivy/graphics/vbo.so
09-18 00:04:30.928 30527 30527 F DEBUG   :       #05 pc 00000000000073c8  /data/data/<APP>/files/app/_python_bundle/site-packages/kivy/graphics/compiler.so
```

```
requirements = kivy==master, pyjnius, https://github.com/kivymd/KivyMD/archive/master.zip
android.api = 33
android.ndk = 25
```
If I use `kivymd==1.0.2` - all fine

**Comment by T-Dynamos** [comment]
https://discord.com/channels/566880874789076992/566880874789076994/1015860048381083678

This is because of opengl

https://github.com/kivymd/KivyMD/issues/1327

**Comment by markqvist** [comment]
Also seeing this on a Samsung Galaxy Tab S7 running Android 12. Reverting to a version before the elevation/shadow shader was implemented "solves" it. For example, dae1aede5b460b89b34c6900e2a714c2245697a5 seems to work (for my purposes at least).

@T-Dynamos, did you find any workaround or solution to this?

**Comment by T-Dynamos** [comment]
@HeaTTheatR I see no defination of smoothstep function in elevation.frag https://github.com/kivymd/KivyMD/blob/master/kivymd/data/glsl/elevation/elevation.frag

Which is used in line

https://github.com/kivymd/KivyMD/blob/c59012117c8801c0cc9b75f05630884cef0c10a8/kivymd/data/glsl/elevation/elevation.frag#L24-L42

After adding 
```
float smoothstep(float a, float b, float x) {
    float t = clamp((x - a) / (b - a), 0.0, 1.0);

    return t * t * (3.0 - (2.0 * t));
}
```
It works but,
Running this code 
```
from kivy.lang import Builder
from kivymd.app import MDApp

KV = """
MDRelativeLayout:
	md_bg_color:0,0,0,1
	MDCard:
		pos_hint:{"center_x":0.5,"center_y":0.5}
		md_bg_color:1,0,0,1
		size_hint:0.5,0.5
		elevation:2
"""

class App(MDApp):
	
	def build(self):
		return Builder.load_string(KV)
		
App().run()
```
Shadow is broken
![Screenshot_2022-10-03-08-30-32-210_ru iiec pydroid3](https://user-images.githubusercontent.com/68729523/193494179-089aa71b-1f16-4a78-be16-83d93bece1dc.jpg)

**Comment by T-Dynamos** [comment]
I found this app to test shaders https://play.google.com/store/apps/details?id=de.markusfisch.android.shadereditor
https://www.shadertoy.com/view/fsdyzB
![Screenshot_2022-10-03-10-27-07-858_de markusfisch android shadereditor](https://user-images.githubusercontent.com/68729523/193504444-1d387fab-50f8-4a58-bcac-52fbbd0268c1.jpg)

**Comment by T-Dynamos** [comment]
I think so problem still persists @Neizvestnyj can you test?

**Comment by Neizvestnyj** [comment]
Yes, problem exists. @HeaTTheatR Reopen issue please.

```
10-14 22:01:31.194 23929 23957 I python  : [INFO   ] [Kivy        ] v2.1.0
10-14 22:01:31.194 23929 23957 I python  : [INFO   ] [Kivy        ] Installed at "/data/user/0/org.test.myapp/files/app/_python_bundle/site-packages/kivy/__init__.pyc"
10-14 22:01:31.194 23929 23957 I python  : [INFO   ] [Python      ] v3.9.9 (main, Oct 14 2022, 21:51:47) 
10-14 22:01:31.194 23929 23957 I python  : [Clang 14.0.6 (https://android.googlesource.com/toolchain/llvm-project 4c603efb
10-14 22:01:31.194 23929 23957 I python  : [INFO   ] [Python      ] Interpreter at ""
10-14 22:01:31.195 23929 23957 I python  : [INFO   ] [Logger      ] Purge log fired. Processing...
10-14 22:01:31.195 23929 23957 I python  : [INFO   ] [Logger      ] Purge finished!
10-14 22:01:31.198 23929 23957 I python  : [INFO   ] [Factory     ] 189 symbols loaded
10-14 22:01:32.072 23929 23957 I python  : [INFO   ] [KivyMD      ] 1.2.0.dev0, git-Unknown, 2022-10-14 (installed at "/data/user/0/org.test.myapp/files/app/_python_bundle/site-packages/kivymd/__init__.pyc")
10-14 22:01:32.106 23929 23957 I python  : [INFO   ] [Image       ] Providers: img_tex, img_dds, img_sdl2 (img_pil, img_ffpyplayer ignored)
10-14 22:01:32.130 23929 23957 I python  : [INFO   ] [Text        ] Provider: sdl2
10-14 22:01:32.159 23929 23957 I python  : [INFO   ] [Window      ] Provider: sdl2
10-14 22:01:32.178 23929 23957 I python  : [INFO   ] [GL          ] Using the "OpenGL ES 2" graphics system
10-14 22:01:32.182 23929 23957 I python  : [INFO   ] [GL          ] Backend used <sdl2>
10-14 22:01:32.182 23929 23957 I python  : [INFO   ] [GL          ] OpenGL version <b'OpenGL ES 3.2 V@415.0 (GIT@033a5b0, I0e419467bc, 1583931100) (Date:03/11/20)'>
10-14 22:01:32.182 23929 23957 I python  : [INFO   ] [GL          ] OpenGL vendor <b'Qualcomm'>
10-14 22:01:32.183 23929 23957 I python  : [INFO   ] [GL          ] OpenGL renderer <b'Adreno (TM) 630'>
10-14 22:01:32.183 23929 23957 I python  : [INFO   ] [GL          ] OpenGL parsed version: 3, 2
10-14 22:01:32.183 23929 23957 I python  : [INFO   ] [GL          ] Texture max size <16384>
10-14 22:01:32.183 23929 23957 I python  : [INFO   ] [GL          ] Texture max units <16>
10-14 22:01:32.213 23929 23957 I python  : [INFO   ] [Window      ] auto add sdl2 input provider
10-14 22:01:32.214 23929 23957 I python  : [INFO   ] [Window      ] virtual keyboard not allowed, single mode, not docked
10-14 22:01:32.272 23929 23957 I python  : 1.2.0.dev0
10-14 22:01:32.395 23929 23957 I python  : [WARNING] [Base        ] Unknown <android> provider
10-14 22:01:32.395 23929 23957 I python  : [INFO   ] [Base        ] Start application main loop
10-14 22:01:32.399 23929 23957 I python  : [INFO   ] [GL          ] NPOT texture support is available
10-14 22:01:32.724 23966 23966 F DEBUG   :       #04 pc 0000000000009e38  /data/data/org.test.myapp/files/app/_python_bundle/site-packages/kivy/graphics/vbo.so
```

**Comment by markqvist** [comment]
Can also confirm the problem persists. On all of the devices I have available to test with, it only happens on one with an Adreno GPU, which matches above cases too.

**Comment by T-Dynamos** [comment]
Now I think so Adding 
```xml
	<uses-feature
		android:glEsVersion="0x00020000"
		>
```
in manifest is the solution

See : https://stackoverflow.com/questions/52801644/android-opengl-app-crashes-when-creating-the-shader

I also confirm that line in this app https://play.google.com/store/apps/details?id=de.markusfisch.android.shadereditor

Because this app able to run same shader
![Screenshot_2022-10-15-08-16-15-543_com gmail heagoo apkeditor pro](https://user-images.githubusercontent.com/68729523/195965551-1af0c0e6-9a0e-4b51-bb1c-e5aca3b76c88.jpg)
![image](https://user-images.githubusercontent.com/68729523/195965570-45f5e5f3-eec9-4089-9ade-470ac2c05377.png)

**Comment by T-Dynamos** [comment]
Can I get an apk for testing?

**Comment by markqvist** [comment]
I already have the GLES version set in AndroidManifest, and the crash still occurs, unfortunately: https://github.com/markqvist/Sideband/blob/main/sbapp/patches/AndroidManifest.tmpl.xml#L25

I am currently disabling the elevation shader completely in my shipped APKs to avoid the issue, so the APKs on my repo does not crash on Adreno GPU. But if you need it for testing, I can build one with shaders enabled for testing.

**Comment by Neizvestnyj** [comment]
@HeaTTheatR Set Bug label (High priority)

**Comment by HeaTTheatR** [comment]
@Neizvestnyj Done!

**Comment by mp-007** [comment]
If someone can test if this bug persist with the new kivy master version (2.2.0), it will be appreciated. I did a PR (https://github.com/kivy/kivy/pull/8043) to update the SDL version to 2.24.1 and this version include some bug fix on Open GL 2 ES compilation.

**Comment by mp-007** [comment]
Update: p4a need to be also updated. I have asked an update in this pull request https://github.com/kivy/python-for-android/pull/2692 to include all bug fix with the open gl 2, but we need to wait the next SDL release.
~~An other thing that I realized is that the SDL version 2.24.1 doesn't changed anything in  Open GL 2 ES compilation, but everything should be add in version  2.25.0 (not release yet)~~

To be continued....

**Comment by HeaTTheatR** [comment]
We can check these PR - https://github.com/kivy/kivy/pull/8056 - that have not yet been merged into the Kivy master branch.  If this is merged and it works well, then we can use this solution in the KivyMD library. However, I was unable to test this PR myself, as I failed to compile Kivy with this PR. 

![Снимок экрана 2022-11-20 в 11 46 49](https://user-images.githubusercontent.com/16930280/202893476-82058fc6-1f59-48c4-a8f2-7224b194952b.png)

**Comment by mp-007** [comment]
for long term solution I think that https://github.com/kivy/kivy/pull/8056 will be a better solution, but it can be good to test the new commit https://github.com/kivy/python-for-android/commit/e14bdfc92f8fd369b21b4f3a8dfd601ea1a3f90d merged into develop branch. It's included several bug fix in opengl es 2.0 from SDL 2.24.1 and 2.26.0 that maybe help to solve our issue.

**Comment by HeaTTheatR** [comment]
@mp-007 I'll try to watch it on the weekend.

**Comment by LazzaAU** [comment]
Can someone please clarify how i get past this issue or are we still waiting on something to be merged / reviewed / fixed ? I'm still un able to compile my app on kivyMD 1.1.1 due to this issue and i''m not clear on what i have to do to make it work. Cheers

**Comment by andmart** [comment]
> Can



> Can someone please clarify how i get past this issue or are we still waiting on something to be merged / reviewed / fixed ? I'm still un able to compile my app on kivyMD 1.1.1 due to this issue and i''m not clear on what i have to do to make it work. Cheers

I am facing the same problem. The AndroidManifest has ``` <uses-feature android:glEsVersion="0x00020000" />``` 
So as @LazzaAU mentioned above, where we are now? 

Thanks in advance.

**Comment by YaROKer** [comment]
Since the question has been open for almost 3 months, I found only one solution for myself, to completely disable shadows by editing "uix/behavior/elevation.py" class CommonElevationBehavior. 
if it helps someone, I attached a file where I commented the lines that were related to the creation of shadows, and tested it by replacing the elevation.py file in master.zip and in buildozer.spec I wrote the path to the zip instead of kivymd, everything works on the commit [7b1e72](https://github.com/kivymd/KivyMD/tree/7b1e72f34c245ce6758d32156e5a7a96842c0579). This permanently disables shadow creation. This is a workaround, not a fix, so we're waiting for a fix.
[elevation.zip](https://github.com/kivymd/KivyMD/files/10226881/elevation.zip)

**Comment by HeaTTheatR** [comment]
I tested the [PR](https://github.com/kivy/kivy/pull/8056)  - and it works. Now I'm refactoring `ElevationBehavior` class according to the changes in this PR. And while this PR has not yet been merged, you can use the `Kivy` installation as follows - `pip install git+https://github.com/DexerBR/kivy.git@boxshadow_instruction` and use `KivyMD` library from master branch.

**Comment by PythonLaboratory** [comment]
@HeaTTheatR, may be a stupid question, but how to write it (`pip install git+https://github.com/DexerBR/kivy.git@boxshadow_instruction`) in buildozer.spec?

**Comment by HeaTTheatR** [comment]
@PythonLaboratory `requirements = git+https://github.com/DexerBR/kivy.git@boxshadow_instruction, https://github.com/kivymd/KivyMD/archive/master.zip`

**Comment by PythonLaboratory** [comment]
@HeaTTheatR, thanks, i will try

**Comment by HeaTTheatR** [comment]
@

> @HeaTTheatR, thanks, i will try

I haven't made any changes to the KivyMD master branch yet. It will be done in two or three hours.

**Comment by PythonLaboratory** [comment]
@HeaTTheatR 
> @PythonLaboratory `requirements = git+https://github.com/DexerBR/kivy.git@boxshadow_instruction, https://github.com/kivymd/KivyMD/archive/master.zip`

when i write like this `requirements = git+https://github.com/DexerBR/kivy.git@boxshadow_instruction, https://github.com/kivymd/KivyMD/archive/master.zip` buildozer throws an error 
```
Updated build/lib.linux-x86_64-cpython-39/kivy/include/config.h

  Updated kivy/include/config.h
  Updated build/lib.linux-x86_64-cpython-39/kivy/include/config.pxi
  Updated kivy/include/config.pxi
  Updated build/lib.linux-x86_64-cpython-39/kivy/setupconfig.py
  Updated kivy/setupconfig.py
  Detected compiler is unix
   error: command '/home/suser/.buildozer/android/platform/android-ndk-r23b/toolchains/llvm/prebuilt/linux-x86_64/bin/clang' failed with exit code 1
  error: subprocess-exited-with-error
 
  × Building wheel for Kivy (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
 
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /home/suser/Projects/TextInputTest/.buildozer/android/platform/build-arm64-v8a/build/venv/bin/python /home/suser/Projects/TextInputTest/.buildozer/android/platform/build-arm64-v8a/build/venv/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py build_wheel /tmp/tmpklha_0ah
  cwd: /tmp/pip-req-build-cfyra79e
  Building wheel for Kivy (pyproject.toml) ... error
  ERROR: Failed building wheel for Kivy
  Running command Building wheel for kivymd (pyproject.toml)
  Ignoring "sys._home = value" override
  fatal: not a git repository (or any of the parent directories): .git
  /tmp/pip-build-env-sgjdyvi5/overlay/lib/python3.9/site-packages/setuptools/config/setupcfg.py:508: SetuptoolsDeprecationWarning: The license_file parameter is deprecated, use license_files instead.
    warnings.warn(msg, warning_class)
``` 

but when I write `pip install git+https://github.com/DexerBR/kivy.git@boxshadow_instruction` all OK

**Comment by HeaTTheatR** [comment]
@PythonLaboratory Yes, I also got this error.

**Comment by PythonLaboratory** [comment]
@HeaTTheatR, how fix it?

**Comment by HeaTTheatR** [comment]
@PythonLaboratory I have no idea.

**Comment by PythonLaboratory** [comment]
@HeaTTheat, sadness...

**Comment by T-Dynamos** [comment]
When we add `requirement=https:/..kivy...GitHub..link`
Bulldozer does not use its default kivy build recipe. To fix it, add custom p4a repo then change kivy download url then specify your fork in buildozer.spec file by using `p4a.source_dir`
See : https://buildozer.readthedocs.io/en/latest/contribute.html?highlight=Fork#write-your-own-recipe

**Comment by PythonLaboratory** [comment]
@T-Dynamos, but I still don't understand what I should write in recipe.sh?

**Comment by T-Dynamos** [comment]
Edit this file : https://github.com/kivy/python-for-android/blob/26d5155bb7156032dcc55f97d12c6ce523f9adae/pythonforandroid/recipes/kivy/__init__.py#L26
Change url and then build

**Comment by PythonLaboratory** [comment]
@T-Dynamos, thank you very much!

**Comment by PythonLaboratory** [comment]
@HeaTTheatR, please correct the [Readme.md](https://github.com/kivymd/KivyMD#how-to-fix-a-shader-bug-on-an-android-device), taking into account the comment @T-Dynamos

**Comment by LazzaAU** [comment]
@T-Dynamos, so to dumb down even one step more for us newbies. Are you saying 
requirements in the spec file should read 

```requirements = kivy,https://github.com/kivymd/KivyMD/archive/master.zip```

and line 26 of __init__.py should read 

```url = 'git+https://github.com/DexerBR/kivy.git@boxshadow_instruction'```

or requirements should read

```requirements = git+https://github.com/DexerBR/kivy.git@boxshadow_instruction, https://github.com/kivymd/KivyMD/archive/master.zip'```

and line 26 should read 

url = something else ?????

**Comment by T-Dynamos** [comment]
I don't quite get it what your saying. I am just saying that if you specify kivy url in  buildozer.spec instead of just writing kivy, it would not use default kivy build recipe. In order to fix it follow these steps
1. Fork this repo https://github.com/kivy/python-for-android/fork
2. Edit pythonforandroid/recipes/kivy/__init__.py#L26 in your fork (change url to your url)
3. Then edit this line in your buildozer.spec https://github.com/kivy/buildozer/blob/a7f53d33f481f17ba031f16d725190f8cb6b3768/buildozer/default.spec#L315 change `p4a.fork=kivy` to your GitHub username. This way it would use p4a fork from your GitHub
4. Try Build it!

Also just keep kivy in buildozer.spec requirements as your using custom url

**Comment by mp-007** [comment]
ShadowBox PR is now merged in kivy master branch https://github.com/kivy/kivy/commit/a7c66880270a93821e1f8ecd613409f008fd2ce8
